simulator/src/scrypto/cmd_coverage.rs

```diff
@@ -0,0 +1,251 @@
+                if nightly { " --toolchain=nightly" } else { "" }
+            );
+            Err(Error::CoverageError(CoverageError::MissingWasm32Target))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn check_rustc_version() -> (bool, String) {
+        let output = Command::new("rustc")
+            .args(&["--version", "--verbose"])
+            .output()
+            .expect("Failed to execute rustc command");
+
+        let output_str = String::from_utf8(output.stdout).expect("Failed to read rustc output");
+        let is_nightly = output_str.contains("nightly");
+        let llvm_major_version = Regex::new(r"LLVM version: (\d+)")
+            .unwrap()
+            .captures(&output_str)
+            .and_then(|cap| cap.get(1).map(|m| m.as_str()))
+            .map(String::from)
+            .unwrap();
+
+        (is_nightly, llvm_major_version)
+    }
+
+    fn check_command_availability(command: String) -> Result<(), Error> {
+        if Command::new(&command).arg("--version").output().is_err() {
+            eprintln!("Missing command: {}. Please install LLVM version matching rustc LLVM version, which is {}.", 
+                command, command.split('-').last().unwrap_or("Unknown"));
+            eprintln!("For more information, check https://apt.llvm.org/");
+            Err(Error::CoverageError(CoverageError::MissingLLVM))
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn run(&self) -> Result<(), Error> {
+        // Verify rust version and wasm target
+        Self::check_wasm_target(false)?;
+
+        let (mut is_nightly, mut llvm_major_version) = Self::check_rustc_version();
+        let mut unset_rustup_toolchain = false;
+        if !is_nightly {
+            // Try to use nightly toolchain automatically
+            env::set_var("RUSTUP_TOOLCHAIN", "nightly");
+            (is_nightly, llvm_major_version) = Self::check_rustc_version();
+            if !is_nightly {
+                eprintln!("Coverage tool requries nightly version of rust toolchain");
+                eprintln!("You can install it by using the following commands:");
+                eprintln!("rustup target add wasm32-unknown-unknown --toolchain=nightly");
+                return Err(Error::CoverageError(CoverageError::IncorrectRustVersion));
+            }
+            Self::check_wasm_target(true)?;
+            unset_rustup_toolchain = true;
+        }
+
+        // Verify that all llvm tools required to generate coverage report are installed
+        Self::check_command_availability(format!("clang-{}", llvm_major_version))?;
+        Self::check_command_availability(format!("llvm-cov-{}", llvm_major_version))?;
+        Self::check_command_availability(format!("llvm-profdata-{}", llvm_major_version))?;
+
+        // Build package
+        let path = self.path.clone().unwrap_or(current_dir().unwrap());
+        let (wasm_path, _) = build_package(&path, false, false, true, Level::Trace, true)
```

Comment:
@iamyulong This approach, with building package before running tests and then reusing it makes tests way faster. Here's my recommendation regarding `scrypto test`:
1. Build all packages in advance, including dependencies (this may be hard when it comes to dependencies)
2. Store information what has been build in ENV, you can also store checksum and build options
3. In `test_runner.rs` try to use cached version, if it doesn't exist then build it and eventually cache it somehow (this may be hard because requires synchronization)

Reply:
This is generally a good idea.  :+1: 

I believe the `cargo build` automatically caches previous builds and avoids re-compiling if not updates have been made, right? However, we can potentially speed up tests by caching package publishing processing and genesis bootstrap. 

This is definitely something that we need to measure and improve.

Reply:
Yeah it does, but it's slow and requires other tests to wait when using multiple workers/threads